src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -256,308 +358,569 @@ EmbeddedAssemblies::add_type_mapping (TypeMappingInfo **info, const char *source
-						(const mono_byte*)map_info.area,
-				        static_cast<int>(info.uncompressed_size)))
-					continue;
-			}
+			default:
+				log_fatal (LOG_DEFAULT, "Internal error: unsupported file type %f for immediate mapping", type);
+				return false;
+		}
+#endif
+	}
+
+	return true;
+}
 
-			if (utils.ends_with (cur_entry_name, ".config") && bundled_assemblies != nullptr) {
-				char *assembly_name = strdup (basename (cur_entry_name));
-				// Remove '.config' suffix
-				*strrchr (assembly_name, '.') = '\0';
+bool
+EmbeddedAssemblies::zip_read_cd_info (int fd, uint32_t& cd_offset, uint32_t& cd_size, uint16_t& cd_entries)
+{
+	// The simplest case - no file comment
+	off_t ret = ::lseek (fd, -ZIP_EOCD_LEN, SEEK_END);
+	if (ret < 0) {
+		log_fatal (LOG_DEFAULT, "Unable to seek into the APK to find ECOD: %s", std::strerror (errno));
+		return false;
+	}
 
-				md_mmap_info map_info = md_mmap_apk_file(fd, offset, info.uncompressed_size, cur_entry_name, apk);
-				mono_register_config_for_assembly (assembly_name, (const char*)map_info.area);
+	uint8_t eocd[ZIP_EOCD_LEN];
+	ssize_t nread = ::read (fd, eocd, static_cast<size_t>(ZIP_EOCD_LEN));
+	if (nread < 0 || nread != ZIP_EOCD_LEN) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD from the APK: %s", std::strerror (errno));
+		return false;
+	}
 
-				continue;
-			}
+	size_t index = 0; // signature
+	uint8_t signature[4];
 
-			if (!(utils.ends_with (cur_entry_name, ".dll") || utils.ends_with (cur_entry_name, ".exe")))
-				continue;
+	if (!zip_read_field (eocd, ZIP_EOCD_LEN, index, signature)) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD signature");
+		return false;
+	}
 
-			if (entry_is_overridden)
-				continue;
+	if (memcmp (signature, ZIP_EOCD_MAGIC, sizeof(signature)) == 0) {
+		return zip_extract_cd_info (eocd, ZIP_EOCD_LEN, cd_offset, cd_size, cd_entries);
+	}
+
+	// Most probably a ZIP with comment
+	size_t alloc_size = 65535 + ZIP_EOCD_LEN; // 64k is the biggest comment size allowed
+	ret = lseek (fd, static_cast<off_t>(-alloc_size), SEEK_END);
+	if (ret < 0) {
+		log_fatal (LOG_DEFAULT, "Unable to seek into the file to find ECOD before APK comment: ", std::strerror (errno));
+		return false;
+	}
+
+	auto buf = new uint8_t[alloc_size];
+	nread = ::read (fd, buf, alloc_size);
+
+	if (nread < 0 || static_cast<size_t>(nread) != alloc_size) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD and comment from the APK: ", std::strerror (errno));
+		return false;
+	}
 
-			size_t alloc_size = MULTIPLY_WITH_OVERFLOW_CHECK (size_t, sizeof(void*), bundled_assemblies_count + 1);
-			bundled_assemblies = reinterpret_cast<MonoBundledAssembly**> (utils.xrealloc (bundled_assemblies, alloc_size));
-			cur = bundled_assemblies [bundled_assemblies_count] = reinterpret_cast<MonoBundledAssembly*> (utils.xcalloc (1, sizeof (MonoBundledAssembly)));
-			++bundled_assemblies_count;
+	// We scan from the end to save time
+	bool found = false;
+	for (size_t i = alloc_size - (ZIP_EOCD_LEN + 2); i >= 0; i--) {
```

Comment:
This loop *may* be why many of our unit tests are timing out, and is *also* a perfect example of how [unsigned integers are error prone](https://github.com/xamarin/xamarin-android/pull/3606/files#diff-fcb50a6533eb26323059bc5c8bda1dd7R687).

Consider!

```c
#include <stdio.h>

int main ()
{
	size_t len = 5;
	for (size_t i = len; i >= 0; i--) {
		printf ("len=%zu\n", i);
	}
	return 0;
}
```

This is (more or less) the same loop construct: an (unsigned) `size_t` iterator type, counting downward.

The result?  Infinite loop:

```
$  cc loop.c
$ ./a.out > o.txt
^C
$ head -8 o.txt
len=5
len=4
len=3
len=2
len=1
len=0
len=18446744073709551615
len=18446744073709551614
...
```

I don't know *for certain* that this loop is causing the hangs, but it certainly *could* be causing the hangs.

Reply:
`i` should have been a `ssize_t`, good catch. This code, however, is for ZIP archives with comments, so unless the APK has one, it will never run.